src/compiler/transformers/module/module.ts

```diff
@@ -604,16 +626,91 @@ namespace ts {
             if (destructuringNeedsFlattening(node.left)) {
-                return flattenDestructuringAssignment(node, moduleExpressionElementVisitor, context, FlattenLevel.All, /*needsValue*/ false, createAllExportExpressions);
+                return flattenDestructuringAssignment(node, visitor, context, FlattenLevel.All, !valueIsDiscarded, createAllExportExpressions);
             }
-            return visitEachChild(node, moduleExpressionElementVisitor, context);
+            return visitEachChild(node, visitor, context);
+        }
+
+        function visitForStatement(node: ForStatement) {
+            return factory.updateForStatement(
+                node,
+                visitNode(node.initializer, discardedValueVisitor, isForInitializer),
+                visitNode(node.condition, visitor, isExpression),
+                visitNode(node.incrementor, discardedValueVisitor, isExpression),
+                visitIterationBody(node.statement, visitor, context)
+            );
+        }
+
+        function visitExpressionStatement(node: ExpressionStatement) {
+            return factory.updateExpressionStatement(
+                node,
+                visitNode(node.expression, discardedValueVisitor, isExpression)
+            );
+        }
+
+        function visitParenthesizedExpression(node: ParenthesizedExpression, valueIsDiscarded: boolean) {
+            return factory.updateParenthesizedExpression(node, visitNode(node.expression, valueIsDiscarded ? discardedValueVisitor : visitor, isExpression));
+        }
+
+        function visitPartiallyEmittedExpression(node: PartiallyEmittedExpression, valueIsDiscarded: boolean) {
+            return factory.updatePartiallyEmittedExpression(node, visitNode(node.expression, valueIsDiscarded ? discardedValueVisitor : visitor, isExpression));
+        }
+
+        function visitPreOrPostfixUnaryExpression(node: PrefixUnaryExpression | PostfixUnaryExpression, valueIsDiscarded: boolean) {
+            // When we see a prefix or postfix increment expression whose operand is an exported
+            // symbol, we should ensure all exports of that symbol are updated with the correct
+            // value.
+            //
+            // - We do not transform generated identifiers for any reason.
+            // - We do not transform identifiers tagged with the LocalName flag.
+            // - We do not transform identifiers that were originally the name of an enum or
+            //   namespace due to how they are transformed in TypeScript.
+            // - We only transform identifiers that are exported at the top level.
+            if ((node.operator === SyntaxKind.PlusPlusToken || node.operator === SyntaxKind.MinusMinusToken)
+                && isIdentifier(node.operand)
+                && !isGeneratedIdentifier(node.operand)
+                && !isLocalName(node.operand)
+                && !isDeclarationNameOfEnumOrNamespace(node.operand)) {

```

Comment:
why do you have to redo this check if the factory sets TransformFlags.ContainsUpdateExpressionForIdentifier?

Reply:
Because not every `PrefixUnaryExpression` or `PostfixUnaryExpression` is one we should transform, and the `TransformFlags.ContainsUpdateExpressionForIdentifier` flag could be set deeper in the parse tree:

```ts
let x;
({ get y() { return x++; }, set y(value) { x = value; }).y++;
export { x }
```
The `TransformFlags.ContainsUpdateExpressionForIdentifier` flag will be set on `x++` and every node that contains it, including the outer `({ ... }).y++`.

Reply:
Many of the `Contains*` flags bubble up to the `SourceFile` so that we can accept an entire subtree if there's no work to be done. The few that don't are due to the specific needs of a transformation.